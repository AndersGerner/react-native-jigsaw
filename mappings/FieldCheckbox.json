[{"name":"Field Checkbox","tag":"FieldCheckbox","description":"A row with left aligned checkbox and body","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Field_Checkbox.png","category":"field","supports_list_render":true,"props":{"title":{"label":"Title","description":"Text to display","type":"string","value":"Beautiful West Coast Villa","editable":true},"color":{"label":"Color","description":"Custom color for the checkbox","editable":true,"required":false,"type":"color"},"disabled":{"label":"Disabled","description":"Whether checkbox and headline is disabled","editable":true,"required":false,"value":false,"type":"boolean"},"fieldName":{"label":"Field name","description":"The name of the field within the screen that will store this component's value","type":"fieldName","value":"checkboxValue","valuePropName":"status","handlerPropName":"onPress","editable":true,"required":false}},"layout":{}}]