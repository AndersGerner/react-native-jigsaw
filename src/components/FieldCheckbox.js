/* @flow */
import * as React from "react";
import { withTheme } from "../core/theming";
import { COMPONENT_TYPES, FORM_TYPES } from "../core/component-types";
import { StyleSheet, View, Text } from "react-native";
import Touchable from "./Touchable";
import Checkbox from "./Checkbox";
import type { Theme } from "../types";
import color from "color";

type Props = {|
  title: string,
  style?: any,
  theme: Theme,
  status: "checked" | "unchecked" | "indeterminate",
  disabled?: boolean,
  color?: string,
  onPress?: () => mixed
|};

class FieldCheckbox extends React.Component<Props> {
  onPress = () => {
    const { onPress } = this.props;
    if (onPress) {
      onPress();
    }
  };

  render() {
    const {
      title,
      status,
      onPress,
      color: checkboxColor,
      disabled,
      theme: { colors, typography, spacing, disabledOpacity }
    } = this.props;

    let titleColor = status === "checked" ? colors.medium : colors.light;

    if (disabled) {
      titleColor = color(titleColor)
        .alpha(disabledOpacity)
        .rgb()
        .string();
    }

    return (
      <Touchable onPress={this.onPress} disabled={disabled}>
        <View style={styles.container}>
          <Checkbox
            status={status}
            disabled={disabled}
            onPress={onPress}
            color={checkboxColor}
          />
          <Text
            style={[
              typography.body1,
              { marginLeft: spacing.medium, color: titleColor }
            ]}
          >
            {title}
          </Text>
        </View>
      </Touchable>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center"
  }
});

export default withTheme(FieldCheckbox);

export const SEED_DATA = [
  {
    name: "Field Checkbox",
    tag: "FieldCheckbox",
    description: "A row with left aligned checkbox and body",
    preview_image_url:
      "https://res.cloudinary.com/altos/image/upload/v1541096663/draftbit/library/jigsaw-1.0/reps/Field_Checkbox.png",
    category: COMPONENT_TYPES.field,
    supports_list_render: true,
    props: {
      title: {
        label: "Title",
        description: "Text to display",
        type: FORM_TYPES.string,
        value: "Beautiful West Coast Villa",
        editable: true
      },
      status: {
        label: "Status",
        description: "Status of checkbox",
        required: true,
        editable: true,
        value: "checked",
        type: FORM_TYPES.flatArray,
        options: ["checked", "unchecked", "indeterminate"]
      },
      color: {
        label: "Color",
        description: "Custom color for the checkbox",
        editable: true,
        required: false,
        type: FORM_TYPES.color
      },
      onPress: {
        label: "onPress handler",
        description: "Function that handles checking or unchecking",
        editable: true,
        required: false,
        value: "{this.onPress}",
        type: FORM_TYPES.function
      },
      disabled: {
        label: "Disabled",
        description: "Whether checkbox and headline is disabled",
        editable: true,
        required: false,
        value: false,
        type: FORM_TYPES.boolean
      }
    },
    layout: {
      width: 343,
      height: 59
    }
  }
];
