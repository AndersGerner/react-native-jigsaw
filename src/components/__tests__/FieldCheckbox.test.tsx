import * as React from "react";
import renderer from "react-test-renderer";
import FieldCheckbox from "../FieldCheckbox";

it("renders field checkbox", () => {
  const tree = renderer
    .create(<FieldCheckbox color="red" status="checked" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
