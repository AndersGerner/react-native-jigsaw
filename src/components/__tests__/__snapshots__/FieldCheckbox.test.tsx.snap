// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders field checkbox 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityLiveRegion="polite"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "red",
            "borderColor": "rgba(165, 173, 183, 1)",
            "borderRadius": 2,
            "borderWidth": 0,
            "height": 25,
            "opacity": 1,
            "width": 25,
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "opacity": 1,
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 26,
            },
            Object {
              "color": "rgba(70, 78, 88, 1)",
              "marginLeft": 12,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
